File: plugins/skill-intake/scripts/extract_open_questions.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const MARKERS = [/\?$/, /要確認/, /未定/, /\bTBD\b/i, /わからない/, /検討中/, /後で/];

function walk(obj, prefix = '', acc = []) {
  if (obj == null) return acc;
  if (typeof obj === 'string') {
    if (MARKERS.some((re) => re.test(obj))) acc.push({ path: prefix, text: obj.trim() });
    return acc;
  }
  if (Array.isArray(obj)) { obj.forEach((v, i) => walk(v, `${prefix}[${i}]`, acc)); return acc; }
  if (typeof obj === 'object') {
    for (const k of Object.keys(obj)) walk(obj[k], prefix ? `${prefix}.${k}` : k, acc);
  }
  return acc;
}

function extract(intake) {
  const detected = walk(intake);
  const existing = Array.isArray(intake.open_questions) ? intake.open_questions : [];
  const merged = [...existing];
  for (const d of detected) {
    if (!merged.some((m) => (typeof m === 'string' ? m : m.text) === d.text)) merged.push(d);
  }
  return { count: merged.length, open_questions: merged };
}

function main(argv) {
  const file = argv[2];
  if (!file) { process.stderr.write('usage: extract_open_questions.js <intake.json>\n'); return 2; }
  let data;
  try { data = JSON.parse(fs.readFileSync(path.resolve(file), 'utf8')); }
  catch (e) { process.stderr.write(`input error: ${e.message}\n`); return 2; }
  const r = extract(data);
  process.stdout.write(JSON.stringify(r, null, 2) + '\n');
  return 0;
}

module.exports = { extract };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/extract_open_questions.py
#!/usr/bin/env python3
"""Walk intake.json and surface strings that look like open questions."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

_MARKERS = [
    re.compile(r'\?$'),
    re.compile(r'要確認'),
    re.compile(r'未定'),
    re.compile(r'\bTBD\b', re.IGNORECASE),
    re.compile(r'わからない'),
    re.compile(r'検討中'),
    re.compile(r'後で'),
]


def walk(obj: Any, prefix: str = '', acc: list[dict[str, str]] | None = None) -> list[dict[str, str]]:
    if acc is None:
        acc = []
    if obj is None:
        return acc
    if isinstance(obj, str):
        if any(m.search(obj) for m in _MARKERS):
            acc.append({'path': prefix, 'text': obj.strip()})
        return acc
    if isinstance(obj, list):
        for i, v in enumerate(obj):
            walk(v, f'{prefix}[{i}]', acc)
        return acc
    if isinstance(obj, dict):
        for k, v in obj.items():
            walk(v, f'{prefix}.{k}' if prefix else k, acc)
    return acc


def extract(intake: dict[str, Any]) -> dict[str, Any]:
    detected = walk(intake)
    existing = intake.get('open_questions') if isinstance(intake.get('open_questions'), list) else []
    existing = list(existing or [])
    merged: list[Any] = list(existing)
    for d in detected:
        text = d['text']
        already = any((m if isinstance(m, str) else m.get('text')) == text for m in merged)
        if not already:
            merged.append(d)
    return {'count': len(merged), 'open_questions': merged}


def main(argv: list[str]) -> int:
    if len(argv) < 1:
        sys.stderr.write('usage: extract_open_questions.py <intake.json>\n')
        return 2
    try:
        data = json.loads(Path(argv[0]).resolve().read_text(encoding='utf-8'))
    except Exception as e:
        sys.stderr.write(f'input error: {e}\n')
        return 2
    r = extract(data)
    sys.stdout.write(json.dumps(r, ensure_ascii=False, indent=2) + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
